Document type variations and update images (#5830)
godotengine/godot-docs: tutorials/ui/gui_theme_type_variations.rst

.. _doc_gui_theme_type_variations:

Overview
========

> When designing a user interface there may be times when a control node
> needs to have a different look than what is normally defined by the
> theme. Theme override properties of a node can be used. However if
> this change needs to be done multiple times the same work is being done
> multiple times, it also requires finding and editing every individual
> node with that look if it needs to be adjusted.

Reviewer: ```suggestion
When designing a user interface there may be times when a :ref:`Control <class_Control>` node
needs to have a different look than what is normally defined by a :ref:`Theme <class_Theme>`.
Every control node has theme property overrides, which allow you to redefine the styling for
each individual UI element.

.. figure:: img/themecheck.png
   :align: center

This approach quickly becomes hard to manage, if you need to share the same custom look
between several controls. Imagine that you use gray, blue, and red variants of :ref:`Button <class_Button>`
throughout your project. Setting it up every time you add a new button element to your interface
is a tedious task.
```

> As a solution to this problem there are theme type variations. These
> are used to create variations of control nodes in a way that is part
> of the theme.

Reviewer: ```suggestion
To help with the organization and to better utilize the :ref:`power of themes <doc_gui_skinning>`
you can use theme type variations. These work like normal theme types, but instead
of being self-sufficient and standalone they extend another, base type.

Following the previous example, your theme can have some styles, colors, and fonts
defined for the ``Button`` type, customizing the looks of every button element in your UI.
To then have a gray, red, or blue button you would create a new type, e.g. ``GrayButton``, and
mark it as a variation of the base ``Button`` type.
```

> These can be used to redefine every aspect of a control node or only one.
> For example, if a button has a defined StyleBoxFlat for its normal state in
> the theme, and a button type variation only has a different font specified,
> any button using that type variation will still use the StyleBoxFlat defined
> for the theme.

Reviewer: ```suggestion
Type variations can replace some aspects of the base type, but keep others.
They can also define properties that the base style hasn't defined. For example,
your ``GrayButton`` can override the ``normal`` style from the base ``Button``
and add ``font_color`` that ``Button`` has never defined. The control will use
a combination of both types giving priority to the type variation.

.. note::
   The way controls resolve what theme items they use from each type and each
   theme is better described in the "Customizing a control" section of :ref:`doc_gui_skinning`.
```

Might also be worth adding an anchor to that other article so we can link to the section directly.

Creating a type variation
-------------------------

To create a type variation open the theme editor, then open the **Manage
Theme Items** window. In the bottom left under **Add Type** enter a name
for the  type variation then press the add button.

Close the manage theme items window. In the theme editor on the right
side click on the **Type** dropdown above the type editor. The type created
in manage theme items should be listed, select it, then switch to the

> tab with a wrench and screwdriver icon.

Reviewer: This is a very long way to add a new type. You can just click the plus icon next to the type selector, enter the name there, and press "Enter". It will create the type and remove the need to select it manually too.

> .. img:: img/base_type.png

Reviewer: ```suggestion
.. figure:: img/base_type.png
   :align: center
```

Using the figure command is better to give it a nicer styling.

> Click on the plus icon. This windows allows you to select the control node
> the type variation is going to be used for. After selecting a node the tabs
> for this type variation should now have the same style options as that control
> node and can be edited.

Reviewer: ```suggestion
Click on the plus icon next to the **Base Type** field. You can select the base type
there, which would typically be the name of a control node class (e.g., ``Button``, ``Label``, etc).
Type variations can also chain and extend other type variations. This works in the
same way control nodes inherit styling of their base class. For example, ``CheckButton``
inherits styles from ``Button`` because corresponding node types extend each other.

After you select the base type, you should now be able to see its properties on the other
tabs in the theme editor. You can edit them as usual.
```

Using a type variation
----------------------

> Now that a type variation has been created it needs to be applied to a
> node. In the inspector, under the theme property of a control node, is
> the **Theme Type Variation** property. To add one click on the pencil icon
> to the left. Then type in the name of the type variation and click the
> check mark or press enter. If a type variation with that name exists it
> will now be used by the node.

Reviewer: ```suggestion
Now that a type variation has been created you can apply it to your nodes.
In the inspector dock, under the **Theme** property of a control node,
you can find the **Theme Type Variation** property. It is empty by default,
which means that only the base type has an effect on this node.

You can either select a type variations from a dropdown list, or input its name
manually. Variations appear on the list only if the type variation belongs to
the project-wide theme, which you can configure in the project settings. For
any other case you have to input the name of the variation manually. Click on
the pencil icon to the right. Then type in the name of the type variation and click the
check mark icon or press enter. If a type variation with that name exists it
will now be used by the node.
```

.. _doc_gui_theme_type_variations:

Overview
========

When designing a user interface there may be times when a :ref:`Control <class_Control>` node
needs to have a different look than what is normally defined by a :ref:`Theme <class_Theme>`.
Every control node has theme property overrides, which allow you to redefine the styling for
each individual UI element.

> .. figure:: img/themecheck.png
>    :align: center

Reviewer: ```suggestion

.. figure:: img/themecheck.png
   :align: center

```

between several controls. Imagine that you use gray, blue, and red variants of :ref:`Button <class_Button>`
throughout your project. Setting it up every time you add a new button element to your interface
is a tedious task.

To help with the organization and to better utilize the :ref:`power of themes <doc_gui_skinning>`
you can use theme type variations. These work like normal theme types, but instead
of being self-sufficient and standalone they extend another, base type.
Following the previous example, your theme can have some styles, colors, and fonts
defined for the ``Button`` type, customizing the looks of every button element in your UI.
To then have a gray, red, or blue button you would create a new type, e.g. ``GrayButton``, and
mark it as a variation of the base ``Button`` type.

Type variations can replace some aspects of the base type, but keep others.
They can also define properties that the base style hasn't defined. For example,
your ``GrayButton`` can override the ``normal`` style from the base ``Button``
and add ``font_color`` that ``Button`` has never defined. The control will use
a combination of both types giving priority to the type variation.

> .. note::
>    The way controls resolve what theme items they use from each type and each
>    theme is better described in the "Customizing a control" section of :ref:`doc_gui_skinning`.

Reviewer: ```suggestion

.. note::
   The way controls resolve what theme items they use from each type and each
   theme is better described in the "Customizing a control" section of :ref:`doc_gui_skinning`.
```

Creating a type variation
-------------------------

To create a type variation open the theme editor, then click the plus icon
next to the **Type** dropdown on the right side of the editor. Type in what
you want to name the theme type variation in the text box then click **Add Type**.
Below the type dropdown are the property tabs, switch to the tab with a wrench
and screwdriver icon.

.. figure:: img/base_type.png
   :align: center

Click on the plus icon next to the **Base Type** field. You can select the base type
there, which would typically be the name of a control node class (e.g., ``Button``, ``Label``, etc).
Type variations can also chain and extend other type variations. This works in the
same way control nodes inherit styling of their base class. For example, ``CheckButton``
inherits styles from ``Button`` because corresponding node types extend each other.

> After you select the base type, you should now be able to see its properties on the other
> tabs in the theme editor. You can edit them as usual.

Reviewer: ```suggestion

After you select the base type, you should now be able to see its properties on the other
tabs in the theme editor. You can edit them as usual.
```

Using a type variation
----------------------

Now that a type variation has been created you can apply it to your nodes.
In the inspector dock, under the **Theme** property of a control node,
you can find the **Theme Type Variation** property. It is empty by default,
which means that only the base type has an effect on this node.

> You can either select a type variations from a dropdown list, or input its name
> manually. Variations appear on the list only if the type variation belongs to

Reviewer: ```suggestion

You can either select a type variations from a dropdown list, or input its name
manually. Variations appear on the list only if the type variation belongs to
```
